### ISSUE TITLE
Kickstart: production configuration updates

### ISSUE THREAD

**Issue by soullivaneuh** [FusionAuth/fusionauth-issues#1379]
### Context

Originally asked by @trollr
in https://github.com/FusionAuth/fusionauth-issues/issues/170,
a way to automate the FusionAuth instance setup was born: [Kickstart](https://fusionauth.io/docs/v1/tech/installation-guide/kickstart/).

A few times before that, I was working on a CI setup automation of FusionAuth for:

1. Production deployment
1. Staging deployment
1. Merge Request review deployment

The main goal is to be able to manage all our override
(settings templates, mail templates, translations...) by files, versioned by git.
It give us a finer control of the evolution of the instance.

Kickstart was not existing yet and I proposed an another approach
in https://github.com/FusionAuth/fusionauth-issues/issues/430
to make FusionAuth more "Configuration friendly".

Waiting a solution, I made a custom setup bash script,
injecting a setup API key and default settings by SQL at start
to be able to manage most of the configuration with API calls.

This is working, but has any custom setup script, we are a bit "out of the scope"
of the standard FusionAuth install and it sometimes lead to hard to debug issues.

This is why I recently migrated from my custom script from Kickstart, and also to try to escape from the
https://github.com/FusionAuth/fusionauth-issues/issues/1266 bug.
Good news! I was able to manage all my needs with this file with so much more simplicity, and less technical depth.

Great feature then, thanks for that!
Unfortunately, I am currently not able to use it because of only one but critical missing part.

### The problematic

The current FusionAuth Kickstart usage has one weakness: It does not handle the production update.

Indeed, according to the [documentation](https://fusionauth.io/docs/v1/tech/installation-guide/kickstart/):

> Kickstart will only run if no API keys, users or tenants exist, so you can safely leave this environment variable configured even after the system has been configured.

It means any already deployed FusionAuth instance will not be updated by the Kickstart configuration file.

A workaround would be to reset the database on each deployment, but this is **not acceptable for production** and will produce unnecessary db reset noises on the staging and review environments (we do not always want to reset our staging instance).

Because of that, we can not use your Kickstart solution.

### The proposition

The easiest way would be to authorize a manual kickstart loading on instance restart.

On that case, we MUST provide a different kickstart file, because the API request has to be adapted (`POST` vs `PATCH/PUT`).

But even if we do that, I am afraid of the lot of common logic duplication just for an API method change...

The more reliable way would be to get rid of the API part from the kickstart file to transform it onto a "raw" configuration file.

I am of course open to discussion and may provide to you more details about our current setup as needed.
A read-only access to our setup repository may also be considered.

**Comment by robotdan** [comment]
Thanks for your detailed and thoughtful description @soullivaneuh - much appreciated. We always appreciate hearing from you as you provide us such great feedback. 

If I understand the issue correctly, you essentially want to a way to use Kickstart for modifications to a running system. 

For example:
1. Use Kickstart to bootstrap the system. This works today. 
2. On reboot, use a different kickstart, to update, remove or add to the current production configuration. 

Step 2 is what does not work today. And it sounds like ideally, if you didn't have to provide a different kickstart file, but we could adapt an existing kickstart to take `POST` request and make it a `PUT` that could work. 

So in theory, something like this:
1. On startup, not eligible for "normal" kickstart
2. If configured to allow modification to a running system continue. 
2. Skip API stuff from Kickstart, this will only run on a pristine system.
3. Perform each API in order, just like we normally would. 
4. If "thing" already exists by unique Id, retry with a `PUT` 
5. Allow `DELETE` requests (which we don't today), and just ignore these in the kickstart file when run on a pristine system.

In theory - does that workflow sound like what you want? This would likely mean that each API in the kickstart specifies an Id on the `POST` request so it could be used to identify if that "thing" by Id already exists for the conversion of the request to `PUT`. 

A workaround, perhaps less elegant, but you could also manage changes like this via a scripted set of APIs that call `PUT` , `PATCH` or `DELETE`.

**Comment by mooreds** [comment]
Would official terraform support also solve this (so you wouldn't use kickstart, but rather terraform)?

https://github.com/FusionAuth/fusionauth-issues/issues/1089

**Comment by soullivaneuh** [comment]
> In theory - does that workflow sound like what you want?

It looks like, yes. :+1:  But I am unsure of one part: What do you mean by a "pristine system"?

> This would likely mean that each API in the kickstart specifies an Id on the POST request so it could be used to identify if that "thing" by Id already exists for the conversion of the request to PUT.

I am not sure that is REST compliant (not an expert), but it could a indeed a solution as it allows me to keep the `POST` usage without any other change.

> A workaround, perhaps less elegant, but you could also manage changes like this via a scripted set of APIs that call

Well, this is what we have since the first implementation of your app. :upside_down_face: 

We have a shell script that does:

1. A raw SQL execution to create a temporary api key and the first admin user (if not exist)
2. Multiple API calls with the temporary key to setup OR update the settings (templates, email templates and some tenant settings)
3. A raw SQL execution to remove the temporary api key

> Would official terraform support also solve this (so you wouldn't use kickstart, but rather terraform)?

We currently does not use this technology, but if it allows to manage settings update on each deployment, then we may consider it.

Hope those details help a bit.

**Comment by kevcube** [comment]
@mooreds terraform support is certainly nice to have, but doesn't work well to translate between local-dev and cloud-deploys.

Because terraform is meant to manage state, but local devs want to blow away their environment and start fresh, managing FA config for local-dev in TF can be annoying. not to mention that we need to teach devs how to use terraform, albeit they only need to know how to scratch the surface.

Because of the above, Kickstart is the better way for us to manage config of local environments by far.

But, kickstart does not translate well to the cloud. If running a containerized deployment, we need a filesystem on which we can place the kickstart file, which is an infrastructural pain point. And when we want to make a configuration change after first creating our cluster, Kickstart just plain doesn't cut it. Therefore Terraform is the best option for in the cloud.

A way to make configuration updates in the kickstart file can be a _good_ solution, but not perfect (at least for us) because of the filesystem requirement.

What would be awesome (and is maybe a feature request for the TF provider) is a way to interpret Kickstart files by the terraform provider and convert that to resources... This would enable us to maintain one config source of truth (kickstart) while still being able to deploy ongoing changes because of terraform's state management.

**Comment by mooreds** [comment]
@kevcube Seems like what you want is a kickstart API where you can feed arbitrary kickstart json to an endpoint and have that executed against the instance. Then that could be wrapped in a terraform provider or mechanism.

Does that make sense?

**Comment by kevcube** [comment]
@mooreds yes both of those would be good! What makes the proposed solution ideal is if we can apply a kickstart more than once, after we've made some updates to it.

**Comment by kevcube** [comment]
@mooreds are there any changes to configuration systems in the pipeline? something like configuration migrations, file-based configuration for the whole system, etc? this is becoming a major pain point between cloud deploys and developer machines: kickstart doesn't work well in both places, and terraform also doesn't work well in both places.

**Comment by mooreds** [comment]
Hi @kevcube ,

Sorry to hear that this is a pain point, and I appreciate you checking in and giving your feedback. At this time, there's no concrete update on this issue. A few options:

* The community provided terraform provider continues to get patches from the community and contributing to that might be the path forward with the most control for you.
* We have some plans to make kickstart more general, but no time commitment for delivery. 
* Other teams have used a migration based system and the FusionAuth APIs to keep systems in sync. Here's a simple example of using rails migrations to do so: https://github.com/FusionAuth/fusionauth-example-rails-lob-integration/tree/master/db

You can see our roadmap for the next few releases here: 

https://github.com/fusionauth/fusionauth-issues/issues?q=is%3Aopen+is%3Aissue+milestone%3A1.37.0

https://github.com/fusionauth/fusionauth-issues/issues?q=is%3Aopen+is%3Aissue+milestone%3A1.38.0

and our general roadmap guidance is here: https://fusionauth.io/docs/v1/tech/core-concepts/roadmap
